File: Count the paths - GFG/count-the-paths.py
#User function Template for python3

from collections import deque
class Solution:
    def possible_paths(self, edges, n, s, d):
        ways=[0 for i in range(n)]
        ways[s]=1
        graph=[[] for i in range(n)]
        for i,j in edges:
            graph[i].append(j)
        q=deque([s])
        while q:
            a=q.popleft()
            for i in graph[a]:
                q.append(i)
                ways[i]+=1
        return ways[d]

#{ 
#  Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
	T=int(input())
	for i in range(T):
		n, m, s, d = input().split()
		n = int(n); m = int(m); s = int(s); d = int(d);
		edges = []
		for _ in range(m):
		    x,y = input().split()
		    x = int(x); y = int(y);
		    edges.append([x,y])
		obj = Solution()
		ans = obj.possible_paths(edges, n, s, d)
		print(ans)


# } Driver Code Ends

File: Count the paths - GFG/count-the-paths.cpp
// { Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

 // } Driver Code Ends
class Solution {
public:

    void dfs(vector<vector<int>>adj,int src,int des,int &cnt,vector<bool>&vis){
        if(src==des){
            cnt++;
            return;
        }
        vis[src]=true;
        for(auto i:adj[src]){
            if(!vis[i]){
                dfs(adj,i,des,cnt,vis);
            }
        }
        vis[src]=false;
        
    }

	int possible_paths(vector<vector<int>>edges, int n, int s, int d){
	    
	    vector<vector<int>>adj(n,vector<int>());
	    for(auto it:edges){
	        adj[it[0]].push_back(it[1]);
	    }
	    vector<bool>vis(n,false);
	    int cnt=0;
	    dfs(adj,s,d,cnt,vis);
	    return cnt;
	}
};

// { Driver Code Starts.
int main(){
	int tc;
	cin >> tc;
	while(tc--){
		int n, m, s, d;
		cin >> n >> m >> s >> d;
		vector<vector<int>>edges;
		for(int i = 0; i < m; i++){
			int u, v;
			cin >> u >> v;
			edges.push_back({u,v});
		}
		Solution obj;
		int ans = obj.possible_paths(edges, n, s, d);
		cout << ans <<"\n";
	}
	return 0;
}  // } Driver Code Ends